include/ndarray/pybind11.h

```diff
@@ -42,13 +42,131 @@
 #include "ndarray.h"
-#include "ndarray/converter/numpy.h"
-#include "ndarray/converter/ufunctors.h"
-#include "ndarray/converter/Vector.h"
-#include "ndarray/converter/eigen.h"
+#include "ndarray/eigen.h"
+#include "ndarray/converter/PyManager.h"
+
+namespace ndarray {
+
+#if PYBIND11_VERSION_MAJOR == 2 && PYBIND11_VERSION_MINOR <= 1
+using pybind11_np_size_t = size_t;
+#else
+using pybind11_np_size_t = ssize_t;
+#endif
+
+template <typename T, int N, int C>
+struct Pybind11Helper {
+    using Element = typename ndarray::Array<T,N,C>::Element;
+    using Wrapper = pybind11::array_t<typename std::remove_const<Element>::type, 0>;  // 0: no ensurecopy
+    static constexpr bool isConst = std::is_const<Element>::value;
+
+    Pybind11Helper() : isNone(false), wrapper() {}
+
+    bool init(pybind11::handle src) {
+        isNone = src.is_none();
+        if (isNone) {
+            return true;
+        }
+        try {
+            wrapper = pybind11::reinterpret_borrow<Wrapper>(src);
+        } catch (pybind11::error_already_set & err) {
+            return false;
+        }
+        return true;
+    }
+
+    bool check() const {
+        if (!isNone) {
+            return true;
+        }
+        if (!wrapper) {
+            return false;
+        }
+        if (wrapper.ndim() != N) {
+            return false;
+        }
+        if (!isConst && !wrapper.writeable()) {
+            return false;
+        }
+        pybind11_np_size_t const * shape = wrapper.shape();
+        pybind11_np_size_t const * strides = wrapper.strides();
+        pybind11_np_size_t const itemsize = wrapper.itemsize();
+        if (C > 0) {
+            ssize_t requiredStride = itemsize;
+            for (int i = 0; i < C; ++i) {
+                if (strides[N-i-1] != requiredStride) {
+                    return false;
+                }
+                requiredStride *= shape[N-i-1];
+            }
+        } else if (C < 0) {
+            ssize_t requiredStride = itemsize;
+            for (int i = 0; i < -C; ++i) {
+                if (strides[i] != requiredStride) {
+                    return false;
+                }
+                requiredStride *= shape[i];
+            }
+        }
+        return true;
+    }
+
+    ndarray::Array<T,N,C> convert() const {
+        if (isNone) {
+            return ndarray::Array<T,N,C>();
+        }
+        Vector<ndarray::Size,N> nShape;
+        Vector<ndarray::Offset,N> nStrides;
+        pybind11_np_size_t const * pShape = wrapper.shape();
+        pybind11_np_size_t const * pStrides = wrapper.strides();
+        pybind11_np_size_t const itemsize = wrapper.itemsize();
+        for (int i = 0; i < N; ++i) {
+            if (pStrides[i] % itemsize != 0) {
+                PyErr_SetString(
+                    PyExc_TypeError,
+                    "Cannot convert array to C++: strides must be an integer multiple of the element size"
+                );
+                throw pybind11::error_already_set();
```

Comment:
Why not simply throw say: `std::invalid_argument`?

Reply:
I think it's a better user experience for Python users to see built-in Python exceptions.